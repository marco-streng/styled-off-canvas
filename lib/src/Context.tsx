import * as React from "react";

type Action = { type: "toggle" | "close" };
type Dispatch = (action: Action) => void;
type State = { isOpen: boolean };
type OffCanvasProviderProps = { children: React.ReactNode };

const OffCanvasStateContext = React.createContext<
  { state: State; dispatch: Dispatch } | undefined
>(undefined);

const offCanvasReducer = (state: State, action: Action) => {
  switch (action.type) {
    case "toggle": {
      return { isOpen: !state.isOpen };
    }

    case "close": {
      return { isOpen: false };
    }

    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

const OffCanvasProvider = ({ children }: OffCanvasProviderProps) => {
  const [state, dispatch] = React.useReducer(offCanvasReducer, {
    isOpen: false,
  });

  const value = { state, dispatch };

  return (
    <OffCanvasStateContext.Provider value={value}>
      {children}
    </OffCanvasStateContext.Provider>
  );
};

const useOffCanvas = () => {
  const context = React.useContext(OffCanvasStateContext);
  if (context === undefined) {
    throw new Error("useOffCanvas must be used within a OffCanvasProvider");
  }
  return context;
};

export { OffCanvasProvider, useOffCanvas };
