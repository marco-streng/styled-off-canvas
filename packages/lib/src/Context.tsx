import * as React from "react";

type Action = { type: "toggle" | "close" | "open" };

type State = { isOpen: boolean; onClose?: () => void };

type OffCanvasProviderProps = {
  children: React.ReactNode;
  isOpen?: boolean;
  onClose?: () => void;
};

const OffCanvasStateContext = React.createContext<
  | {
      isOpen: boolean;
      onClose?: () => void;
      close: () => void;
      open: () => void;
      toggle: () => void;
    }
  | undefined
>(undefined);

const offCanvasReducer = (state: State, action: Action) => {
  switch (action.type) {
    case "toggle": {
      if (state.isOpen && state.onClose) {
        state.onClose();
      }

      return { ...state, isOpen: !state.isOpen };
    }

    case "open": {
      return { ...state, isOpen: true };
    }

    case "close": {
      if (state.onClose) {
        state.onClose();
      }

      return { ...state, isOpen: false };
    }

    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

const OffCanvasProvider = ({
  children,
  isOpen,
  onClose,
}: OffCanvasProviderProps) => {
  const [state, dispatch] = React.useReducer(offCanvasReducer, {
    isOpen: isOpen ?? false,
    onClose,
  });

  const value = {
    isOpen: state.isOpen,
    onClose: state.onClose,
    close: () => dispatch({ type: "close" }),
    open: () => dispatch({ type: "open" }),
    toggle: () => dispatch({ type: "toggle" }),
  };

  return (
    <OffCanvasStateContext.Provider value={value}>
      {children}
    </OffCanvasStateContext.Provider>
  );
};

const useOffCanvas = () => {
  const context = React.useContext(OffCanvasStateContext);
  if (context === undefined) {
    throw new Error("useOffCanvas must be used within an OffCanvasProvider");
  }
  return context;
};

export { OffCanvasProvider, useOffCanvas };
