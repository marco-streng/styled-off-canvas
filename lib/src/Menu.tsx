import React from "react";
import styled from "styled-components";
import { useOffCanvas } from "./Context";

const Container = styled.div<{
  $isOpen: boolean;
  $background?: string;
  $duration?: string;
  $position?: "left" | "right";
  $width?: string;
}>(
  ({
    $isOpen,
    $background = "#fff",
    $duration = "500ms",
    $position = "right",
    $width = "300px",
  }) => `
  background-color: ${$background};
  height: 100%;
  left: ${$position === "left" ? ($isOpen ? 0 : "-100%") : "inherit"};
  max-width: ${$width};
  overflow-y: auto;
  position: fixed;
  right: ${$position === "right" ? ($isOpen ? 0 : "-100%") : "inherit"};
  top: 0;
  transition: ${$position} ${$duration} ease-in-out;
  width: 100%;
  z-index: 10;
`
);

const isCSR = typeof window !== "undefined";

const initalBodyStyle = {
  maxHeight: isCSR ? window.document.body.style.maxHeight : "none",
  overflow: isCSR ? window.document.body.style.overflow : "visbile",
};

/**
 * Menu component
 *
 * @param   {string}    className styled-components className for custom styling
 * @returns {component}           React component
 */
export const Menu = ({
  children,
  background,
  duration,
  position,
  width,
  closeOnEsc,
  className,
  ...props
}: {
  children: React.ReactNode;
  background?: string;
  duration?: string;
  position?: "left" | "right";
  width?: string;
  closeOnEsc?: boolean;
  className?: string;
}) => {
  const { isOpen, close } = useOffCanvas();

  // Avoid scrolling on content when the navigation is open
  React.useEffect(() => {
    const bodyElStyle = window.document.body.style;

    if (isOpen) {
      bodyElStyle.maxHeight = "100%";
      bodyElStyle.overflow = "hidden";
    } else {
      bodyElStyle.maxHeight = initalBodyStyle.maxHeight;
      bodyElStyle.overflow = initalBodyStyle.overflow;
    }
  }, [isOpen]);

  if (closeOnEsc) {
    React.useEffect(() => {
      const handleKeyDown = (event: KeyboardEvent) => {
        if (event.code.toLowerCase() === "escape") {
          close();
        }
      };

      isOpen
        ? window.document.addEventListener("keydown", handleKeyDown)
        : window.document.removeEventListener("keydown", handleKeyDown);
    }, [isOpen]);
  }

  return (
    <Container
      className={className}
      $background={background}
      $duration={duration}
      $isOpen={isOpen}
      $position={position}
      $width={width}
      {...props}
    >
      {children}
    </Container>
  );
};
