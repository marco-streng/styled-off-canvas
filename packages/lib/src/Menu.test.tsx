import { OffCanvasProvider } from "./Context";
import { Menu } from "./Menu";
import { fireEvent, render } from "./utils/test-utils";

describe("Menu", () => {
  test("closed", () => {
    const { container } = render(
      <OffCanvasProvider>
        <Menu>
          <div>content</div>
        </Menu>
      </OffCanvasProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("initial open", () => {
    const { container } = render(
      <OffCanvasProvider isOpen>
        <Menu>
          <div>content</div>
        </Menu>
      </OffCanvasProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("custom color", () => {
    const { container } = render(
      <OffCanvasProvider isOpen>
        <Menu background="#000">
          <div>content</div>
        </Menu>
      </OffCanvasProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("custom duration", () => {
    const { container } = render(
      <OffCanvasProvider isOpen>
        <Menu duration="1s">
          <div>content</div>
        </Menu>
      </OffCanvasProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("position left", () => {
    const { container } = render(
      <OffCanvasProvider isOpen>
        <Menu position="left">
          <div>content</div>
        </Menu>
      </OffCanvasProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("position left closed", () => {
    const { container } = render(
      <OffCanvasProvider>
        <Menu position="left">
          <div>content</div>
        </Menu>
      </OffCanvasProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("closeOnEsc", () => {
    const cb = vi.fn();
    const { container } = render(
      <OffCanvasProvider isOpen onClose={cb}>
        <Menu closeOnEsc>
          <div>content</div>
        </Menu>
      </OffCanvasProvider>
    );

    const before = (container.firstChild as any).className;

    fireEvent.keyDown(container, {
      key: "Escape",
      code: "Escape",
      keyCode: 27,
      charCode: 27,
    });

    expect(before).not.toBe((container.firstChild as any).className);
  });

  test("closeOnEsc callback", () => {
    const cb = vi.fn();
    const { container } = render(
      <OffCanvasProvider isOpen onClose={cb}>
        <Menu closeOnEsc>
          <div>content</div>
        </Menu>
      </OffCanvasProvider>
    );

    fireEvent.keyDown(container, {
      key: "Escape",
      code: "Escape",
      keyCode: 27,
      charCode: 27,
    });

    expect(cb).toHaveBeenCalled();
  });

  test("closeOnEsc with another key", () => {
    const cb = vi.fn();
    const { container } = render(
      <OffCanvasProvider isOpen onClose={cb}>
        <Menu>
          <div>content</div>
        </Menu>
      </OffCanvasProvider>
    );

    fireEvent.keyDown(container, {
      key: "a",
      code: "KeyA",
      keyCode: 65,
      charCode: 65,
    });

    expect(cb).not.toHaveBeenCalled();
  });

  test("body styling", () => {
    const { container } = render(
      <OffCanvasProvider isOpen>
        <Menu closeOnEsc>
          <div>content</div>
        </Menu>
      </OffCanvasProvider>
    );

    expect(document.body.style.overflow).toBe("hidden");
    expect(document.body.style.maxHeight).toBe("100%");

    // Close
    fireEvent.keyDown(container, {
      key: "Escape",
      code: "Escape",
      keyCode: 27,
      charCode: 27,
    });

    expect(document.body.style.overflow).toBe("");
    expect(document.body.style.maxHeight).toBe("");
  });
});
