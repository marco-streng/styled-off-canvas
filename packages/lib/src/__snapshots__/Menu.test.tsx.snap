// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Menu > closed 1`] = `
<div
  class="sc-bcXGCL hNHwdq"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu > custom color 1`] = `
<div
  class="sc-bcXGCL czsTlN"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu > custom duration 1`] = `
<div
  class="sc-bcXGCL eeJWaS"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu > initial open 1`] = `
<div
  class="sc-bcXGCL hZteQD"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu > position left 1`] = `
<div
  class="sc-bcXGCL dwILLQ"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu > position left closed 1`] = `
<div
  class="sc-bcXGCL dKmNBl"
>
  <div>
    content
  </div>
</div>
`;
